graphql/query_resolver.go

```diff
@@ -865,101 +864,85 @@ func (r *queryResolver) MainlineCommits(ctx context.Context, options MainlineCom
 		}
 	}
 
-	index := 0
+	matchingVersionCount := 0
 	versionsCheckedCount := 0
+	hasFilters := isPopulated(buildVariantOptions) && utility.FromBoolPtr(options.ShouldCollapse)
 
-	// We will loop through each version returned from GetMainlineCommitVersionsWithOptions and see if there is a commit that matches the filter parameters (if any).
+	// We will loop through each version returned from GetMainlineCommitVersionsWithOptions and see if there is a commit
+	// that matches the filter parameters (if any).
 	// If there is a match, we will add it to the array of versions to be returned to the user.
-	// If there are not enough matches to satisfy our limit, we will call GetMainlineCommitVersionsWithOptions again with the next order number to check and repeat the process.
+	// If there are not enough matches to satisfy our limit, we will call GetMainlineCommitVersionsWithOptions again
+	// with the next order number to check and repeat the process.
 	for matchingVersionCount < limit {
-		// If we no longer have any more versions to check break out and return what we have.
+		// If we no longer have any more versions to check, break out of the loop.
 		if len(versions) == 0 {
 			break
 		}
-		// If we have checked more versions than the MaxMainlineCommitVersionLimit then break out and return what we have.
+
+		// If we have checked more versions than the MaxMainlineCommitVersionLimit, break out of the loop.
 		if versionsCheckedCount >= model.MaxMainlineCommitVersionLimit {
-			// Return an error if we did not find any versions that match.
 			if matchingVersionCount == 0 {
 				return nil, ResourceNotFound.Send(ctx, fmt.Sprintf("Matching version not found in %d most recent versions", model.MaxMainlineCommitVersionLimit))
 			}
 			break
 		}
-		versionsCheckedCount++
-		v := versions[index]
-		apiVersion := restModel.APIVersion{}
-		apiVersion.BuildFromService(v)
-
-		// If the version was created before we started caching activation status we must manually verify it and cache that value.
-		if v.Activated == nil {
-			err = setVersionActivationStatus(ctx, &v)
-			if err != nil {
-				return nil, InternalServerError.Send(ctx, fmt.Sprintf("Error setting version activation status: %s", err.Error()))
-			}
-		}
-		mainlineCommitVersion := MainlineCommitVersion{}
-		shouldCollapse := false
-		if !utility.FromBoolPtr(v.Activated) {
-			shouldCollapse = true
-		} else if isPopulated(buildVariantOptions) && utility.FromBoolPtr(options.ShouldCollapse) {
+
+		var hasMatchingTasksMap map[string]bool
```

Comment:
I'd recommend renaming to versionHasMatchingTasksMap (or at least commenting -- it's not immediately clear what the subject of the map is, i.e. what "string" corresponds to)

Reply:
Better yet, it seems like we only need to know what versions _do_ match tasks, not just to have all IDs in the map. I'd suggest just turning this into a versionsMatchingTasksMap (i.e. the concurrentlyBuild function can just not put versions in that have no matching tasks). This is a common usage of "maps as sets"

Reply:
Do you want me to add an if-statement to only have entries of `[id]: true` in the map? :0 (I can definitely do that I just wasn't sure)

Reply:
yeah exactly, i would suggest just not putting versions in that aren't true, to make this data structure behave more like a set and less like a dictionary

Reply:
okay tried to clean up the logic here! Lmk what you think!